R/calcHDDCDD.R

```diff
@@ -0,0 +1,99 @@
+#' HDD and CDD scenario and historic data
+#'
+#' smoothes the HDD and CDD scenario data
+#'
+#' @param tlimit Temperature threshold for computing HDD and CDD: 18, 21 or 25
+#'
+#' @author Antoine Levesque
+#'
+#' @importFrom madrat readSource calcOutput
+#' @importFrom magclass setNames setYears as.magpie getNames lowpass getYears mbind
+#' @importFrom dplyr %>%
+#' @importFrom quitte as.quitte
+#' @importFrom utils tail
+#' @export
+
+calcHDDCDD <- function(tlimit = 18) {
+
+  data <- readSource("HDDCDD", subtype = tlimit)
+
+  # yearsScenario = all years for which history is either NA or 0
+  index <- apply(data[, , paste0("history.HDD.history.", tlimit)], c(2, 3),
+                 function(x) {
+                   return(all(is.na(x) | x == 0))
+                 })
+
+  yearsScenario <- c(gsub("y", "", attributes(index)$dimnames$period[index]) %>% as.integer(), 2100)
+  yearsHistory <- setdiff(getYears(data, as.integer = TRUE), yearsScenario)
+
+  convergenceYear <- 2030
+  convergenceYears <- c(max(yearsHistory), yearsScenario[yearsScenario <= convergenceYear])
+  afterConvergenceYears <- yearsScenario[!yearsScenario %in% convergenceYears]
+  lambda <- c(rep(1, length(yearsHistory)),
+              tail(seq(1, 0, length.out = length(convergenceYears)), -1),
+              rep(0, length(afterConvergenceYears)))
+  names(lambda) <- c(yearsHistory, yearsScenario)
+  lambda <- as.magpie(lambda)
+
+  rcps     <- grep("history", getNames(data, TRUE)$rcp,      value = TRUE, invert = TRUE)
+  ssps     <- grep("history", getNames(data, TRUE)$scenario, value = TRUE, invert = TRUE)
+  variable <- grep("history", getNames(data, TRUE)$variable, value = TRUE, invert = TRUE)
+
+  data[is.na(data)] <- 0
+  data <- mbind(data, (1 / 3) * (setYears(data[, 2099, ], 2100) +
+                                 setYears(data[, 2098, ], 2100) +
+                                 setYears(data[, 2097, ], 2100))) # give a point for 2100
+
+  # distribute the history between all scenarios
+  data <- do.call("mbind", lapply(rcps, function(rcp) {
+    do.call("mbind", lapply(ssps, function(ssp) {
+      tmp <- data[, , rcp][, , ssp]
+      tmp[, yearsHistory, ] <- setNames(data[, yearsHistory, "history"][, , "history"],
+                                        getNames(tmp[, yearsHistory, ]))
+      return(tmp)
+    }))
+  }))
+
+
+  # Smooth the trajectories
+  dataSmoothed <- lapply(ssps, function(scenario) {
+    lapply(rcps, function(rcp) {
+      thirdDim <- paste(scenario, variable, rcp, tlimit, sep = ".")
+      out <- lowpass(data[, , thirdDim], i = 100)
+      return(out)
+    })
+  })
+  dataSmoothed <- do.call("mbind", unlist(dataSmoothed, recursive = FALSE))
```

Comment:
Just a side note: I prefer passing the function instead of its name in situations like this, because RStudio will warn if the function does not exist and resolving a function name (string) to an actual function feels like unnecessary magic to me. That's definitely personal preference though, so feel free to leave it :slightly_smiling_face:  (If you change it I'd suggest changing it everywhere, searching for `do.call` should do it)
```suggestion
  dataSmoothed <- do.call(mbind, unlist(dataSmoothed, recursive = FALSE))
```